Add Somfy MyLink support for Covers and Scenes

bendews opened the pull request:
## Description:
This adds integration for the Somfy MyLink platform based on the Synergy JsonRPC API. It will automatically gather configured blinds and scenes from the MyLink hub and also allows additional options configured on a per-cover basis if required.

- Allows direction reversal
- Implements **optional** emulated position tracking based on time taken to open/close covers.

There is one issue that I would like some support with before release - The state gets reset whenever the component is reloaded. Is there any way to read the previous recorded state and keep it?

Thanks!

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#6988

## Example entry for `configuration.yaml` (if applicable):
```yaml
somfy_mylink:
  host: 10.1.1.100
  system_id: mylink_id
  default_move_time: 21.5
  default_reverse: false
  entity_config:
    cover.bedroom:
        move_time: 20.5
    cover.outdoor_awning:
        reverse: true
```

## Checklist:
  - [X] The code change is tested and works locally.
  - [X] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**

If user exposed functionality or configuration variables are added/changed:
  - [X] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [X] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [X] (N/A) New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [X] (N/A) New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [X] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [X] (N/A) Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

bendews committed:
Initial Somfy MyLink components

bendews committed:
Sorted and removed unnecessary imports

bendews committed:
Added somfy_mylink dependency to Cover and Scene component

bendews committed:
Pass configuration data via the discovery_info parameter to child components

bendews committed:
Fixed miscellaneous formatting issues

bendews committed:
Improved exception handling for hub connection

bendews committed:
Changed default kwarg values for ‘build_message’ function

bendews committed:
Fixed line-length issue for log message

bendews committed:
Updated .coverage.rc

bendews committed:
Adjusted misc formatting

homeassistant commented:
Hi @bendews, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#17654).

Once you do that we will be able to review and accept this pull request.

Thanks!

bendews committed:
Adjust logging levels to be more conservative

bendews committed:
Removed redundant line break

MartinHjelmare commented on homeassistant/components/somfy_mylink.py:
```diff
@@ -0,0 +1,136 @@
+    def send_message(self, message, retry_count=3):
+        """Send a message to MyLink using JsonRPC via Socket."""
+        message_as_bytes = str.encode(json.dumps(message))
+        sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
```
Device specific I/O and data parsing should be moved to a standalone library published on pypi.

MartinHjelmare reviewed (changes requested):
Please extract device specific api code to standalone library on pypi.

bendews committed:
Added PyPi library to requirements

bendews committed:
Added new library to component/platform creation

bendews committed:
Miscellaneous fixes to enable Async

bendews committed:
Organised imports

bendews committed:
Added CodeOwner declaration

bendews committed:
Fix line length issues

bendews committed:
Updated import method and Somfy library version

MartinHjelmare reviewed (commented):
Thanks for the PR!

I don't think the emulated position tracking belongs in home assistant. It could be moved to the library, or not included. @balloob?

See below for more comments.

Regarding loading last state, this can be done using `async_get_last_state` in homeassistant.helpers.restore_state.py. Look at other components that use this for examples. We're currently transitioning to a new system to do this in https://github.com/home-assistant/home-assistant/pull/17270.

bendews commented on homeassistant/components/somfy_mylink.py:
```diff
@@ -0,0 +1,47 @@
+    from somfy_mylink_synergy import SomfyMyLinkSynergy
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN][CONF_PORT]
+    system_id = config[DOMAIN][CONF_PASSWORD]
```
MyLink references it as "System ID" however it is used for authentication and functions more like a password, so much so I didn't think it was worth adding a new config constant just for it. Happy to go either way on your recommendation (reference it as `password` in code or change all references to `system_id`).

MartinHjelmare replied:
If the user will be familiar with "System ID", let's make the config string `'system_id'`. Create a local constant for that in the module.

bendews commented on homeassistant/components/cover/somfy_mylink.py:
```diff
@@ -0,0 +1,206 @@
+        cover_config = dict()
+        cover_config['target_id'] = cover['targetID']
+        for cover_opt in cover_options:
+            if cover_opt.get('name') in ('*', cover['name']):
```
Star is a wildcard. Common use case is where a user has multiple awnings which commonly require reversal of the up/down commands.

Rather than explicitly defining every awning (and needing to update `configuration.yaml` when they change) they can use a wildcard entry to change the default behaviour (I.E `reverse: true`) and then explicitly define the exceptions if required.

I agree it's not a great way to do it but I looked lots of components trying to see how that sort of functionality had been implemented and couldn't find any examples of it. Only other alternative I could think of was to have new config constants like `default_reverse` and `default_move_time` but that is also pretty messy!

bendews committed:
Changed CONF_PASSWORD to CONF_SYSTEM_ID

bendews committed:
Use entity config instead of cover options

bendews committed:
Added guard clause to check discovery_info is not empty

bendews committed:
Added ability to track cover state between restarts

bendews committed:
Created wrapper functions for opening/closing to simplify actions when cover is in reversed state

bendews committed:
Added check for entity ID in entity_config

bendews committed:
Updated to use async functions

bendews committed:
Adjusted imports

bendews committed:
Added guard clause for when discovery info is empty

bendews committed:
Updated list syntax

bendews committed:
Removed inessential function (supported_features)

bendews committed:
Adjusted whitespace

bendews committed:
Sorted imports

bendews committed:
Updated to new state restore method

bendews committed:
Updated state restore syntax to use super()

bendews committed:
Merge branch 'dev' into feature/somfy-mylink

bendews committed:
Test

bendews committed:
Updated requirements_all.txt

bendews committed:
Fixed lint error (imperative docstring language)

bendews committed:
Merge pull request #2 from home-assistant/dev

Update to latest

bendews committed:
Merge branch 'dev' into feature/somfy-mylink

bendews committed:
Remove debug line in README

bendews commented:
@MartinHjelmare - Apologies, the PR got a bit messy trying to get those green ticks. Testing framework is absolutely awesome but got a little finicky at times 😄 

I have taken on your feedback and fixed everything, with exception to the emulated position tracking. I did look at implementing it in the 3rd-party library but ended up down a rabbit hole of re-implementing `async_track_time_change`, I think it makes more sense to keep it within the component.

The config structure has changed now also, I removed `cover_options` and implemented an `entity_config`-type of structure. Default values are now top level arguments as you can see below:

```yaml
somfy_mylink:
  host: 10.1.1.100
  password: mylink_id
  default_move_time: 21.5
  default_reverse: false
  entity_config:
    cover.bedroom:
        move_time: 20.5
    cover.outdoor_awning:
        reverse: true
```

A minimal config such as this will still work to autodiscover and control all the available covers and scenes:

```yaml
somfy_mylink:
  host: 10.1.1.100
  password: mylink_id
```

Thanks!

bendews committed:
Removed logging in component left over from debugging

MartinHjelmare commented on homeassistant/components/cover/somfy_mylink.py:
```diff
@@ -0,0 +1,234 @@
+_LOGGER = logging.getLogger(__name__)
+DEPENDENCIES = ['somfy_mylink']
+DEFAULT_SUPPORTED_FEATURES = (SUPPORT_OPEN | SUPPORT_STOP | SUPPORT_CLOSE)
+ENTITY_ID_FORMAT = 'cover.{}'
```
Import this from the cover component.

MartinHjelmare commented on homeassistant/components/cover/somfy_mylink.py:
```diff
@@ -0,0 +1,234 @@
+    for cover in mylink_status['result']:
+        cover_config = {}
+        cover_config['target_id'] = cover['targetID']
+        for key, val in config_options.get('default', {}).items():
```
Looks like we can just copy the dict returned here. We don't need to iterate it.
```py
new = dict(old)
```

bendews replied:
Classic case of over-thinking things, this is much better as a copy or merge. Thanks!

MartinHjelmare commented on homeassistant/components/cover/somfy_mylink.py:
```diff
@@ -0,0 +1,234 @@
+        if self._state is not None:
+            return
+        state = await super().async_get_last_state()
+        if state:
```
Invert this and make it a guard clause by returning if true.

MartinHjelmare commented on homeassistant/components/cover/somfy_mylink.py:
```diff
@@ -0,0 +1,234 @@
+        """Move the cover to a specific position."""
+        position = kwargs.get(ATTR_POSITION)
+        self._set_position = round(position, -1)
+        if self._position == position:
```
We can never adjust the position if we have the wrong position information if we include this check. We shouldn't do that.

MartinHjelmare commented on homeassistant/components/somfy_mylink.py:
```diff
@@ -0,0 +1,81 @@
+    entity_config = config[DOMAIN].get(CONF_ENTITY_CONFIG, {})
+    for default_key in [CONF_DEFAULT_MOVE_TIME, CONF_DEFAULT_REVERSE]:
+        config_key = default_key.replace('default_', '')
+        if config[DOMAIN].get(default_key, None):
```
`None` is the default value returned if the key is missing in the dict.

MartinHjelmare commented on homeassistant/components/somfy_mylink.py:
```diff
@@ -0,0 +1,81 @@
+    host = config[DOMAIN][CONF_HOST]
+    port = config[DOMAIN][CONF_PORT]
+    system_id = config[DOMAIN][CONF_SYSTEM_ID]
+    entity_config = config[DOMAIN].get(CONF_ENTITY_CONFIG, {})
```
There's already a default in the config schema.

MartinHjelmare commented on homeassistant/components/somfy_mylink.py:
```diff
@@ -0,0 +1,81 @@
+    for default_key in [CONF_DEFAULT_MOVE_TIME, CONF_DEFAULT_REVERSE]:
+        config_key = default_key.replace('default_', '')
+        if config[DOMAIN].get(default_key, None):
+            entity_config.setdefault('default', {})
```
All of the config parsing should be done in the config schema.

bendews replied:
This is loading the default values in to the `discovery_info`, so that those values get applied first when setting up each cover.

Is there a better way to pass that data through to the child component without hijacking `discovery_info`? Accessing `config` within the child component just gives an empty dict

MartinHjelmare replied:
We're updating a config item with default values. That should be done in the config schema.

MartinHjelmare replied:
I think we should do it in the config schema with a custom validator appended to the end so that it accesses all needed items in the passes in config.

MartinHjelmare replied:
Look at the `_state_validator` function in the manual alarm_control_panel platform.

MartinHjelmare reviewed (changes requested):
I still don't think we should include the time tracking of the covers in this PR. We're adding a feature that the integration doesn't have. We shouldn't do that.

You can ask for a second opinion from the core team.

bendews commented:
Thanks @MartinHjelmare.
Where is the best place to ask the core team? Other than summoning @balloob 😄 
My only concern is that without some way of emulating position it's impossible to do any automation to open them to 50 or 80% etc, which is a common task to have to do manually (especially for eastern facing blinds getting the harsh mid-morning sun).
Sure having remote up/down controls is useful but that doesn't really cover the use-case for automating them in a weather/sunlight/time context, and can already do that through the somfy app.

balloob commented:
We shouldn't do simulation of any data. Platforms should just provide the available data and leave it up to the consumers of the services and data to handle this.

A user can easily write an automation that will wait X seconds before calling stop etc.

bendews committed:
Merge pull request #6 from home-assistant/dev

Update Dev

balloob commented:
This PR seems to have gone stale. Closing it.

bendews commented:
@balloob @MartinHjelmare  - Without the state tracking this integration is quite useless for me and plenty of others (and I would prefer to not have to maintain two branches). I completely understand the hesitation as this is not a 'generic' platform so it doesn't make sense for one brand/integration to have this and not the others, although Somfy/Tahoma are some of the only major manufacturers to not offer position tracking out-of-the-box.

Would you be open for me to create some form of "Generic XXX" component? Similar to how the Generic Thermostat ties together a temp sensor and heat/cool switches, this could tie together a dumb cover and add position tracking for those who find it useful. Means it could also work for Garage doors, cheap RF covers etc. as well.

Reason as to why this is so important - Shutters/Awnings/Blinds have different requirements based on their position (NESW-facing), location in the home, weather and season. In my part of the world they are used extensively for climate control, as our houses have little/no eaves, very expensive electricity and very hot days, just having a single blind in the optimal position can save $5 a day in cooling costs.

Just as an example I have (for most blinds in my home, but not all!):
- Set to 80% at midday
    - but only 30% if it is a hot day
    - or to 100% if it is overcast/raining
- Close in late afternoon if above 35c
- Set to 50% in the evening
- Close at night
- During the day I might ask Google Home or Homekit to set certain blinds to a certain percentage for whatever reason
    - 70% to block out neighbors but to keep light
    - 100% if its a nice day
    - List goes on etc.

All of these work as automations in isolation, but trying to track the state between all of these, especially the on-demand changes via google home/homekit would be a nightmare. Then add in that there is a possibility to have 16 of these per hub (someone using my smartthings integration has 3 hubs!), all needing to work in unison and requiring different time values.
It would require 100's if not 1000's of lines of automations and multiple emulated switches to try and track state, just completely not worth the time.
Compared to 27 lines of code to do it all for you (possibly less after refactoring) 😄

balloob commented:
If you want to change how we approach building our integrations, open an issue in the architecture repo. A PR is not the place to change policy.

The policy is that we represent the data that we get from an integration as-is, without enhancing it. Enhancements always end up biting us in the future.

bendews commented:
@balloob, i agree that it should not be included as part of this Somfy component, hence the question of doing something similar to the Generic Thermostat component and offering it as a generic way to add the functionality to other devices which also lack the feature, I don't think theres a need to change the architecture at all.
